### ISSUE TITLE
JSON parsing error: {} converted into 'null'

### ISSUE THREAD

**Issue by jacopoMauro** [jolie/jolie#124]
To convert a JSON string output_string into a JSON value I use the following code:

output_string.strictEncoding = true;
getJsonValue@JsonUtils(output_string)(response);

The problem is that while trying to convert the string '{"dead_features": {}, "false_optionals": {}}' into a JSON value I get in output the following JSON object  {"false_optionals":null,"dead_features":null} even if I require a strict encoding.

It seems that there is a small bug in the json_utils

**Comment by mwallnoefer** [comment]
I have looked into this and it's not so easy to fix. Let me explain:

I hope that you have some knowledge about the internal Jolie value representation (reference: https://docs.jolie-lang.org/documentation/basics/data_structures.html). A Jolie value has a root which may be defined or not (_null_) and then zero or more children which are other Jolie values respectively arrays.

This construction has its limitations: the Jolie's JSON parser is not able to distinguish between _null_ and _{}_, since both are values with an undefined root and no children. So it currently prefers to generate _null_s instead of _{}_ when converting back from a Jolie value into a JSON string.

I attach you the respective code parts for consultation. Maybe you or some other developer come up with a proposal to mitigate this?
Of course we will have to be careful to not break the REST JSON and the JSON-RPC parts.

References:
[1] JsonUtils service: https://github.com/jolie/jolie/blob/master/javaServices/coreJavaServices/src/joliex/util/JsonUtilsService.java
[2] Json parser: https://github.com/jolie/jolie/blob/master/lib/jolie-js/src/jolie/js/JsUtils.java
[3] Json tester (for better understanding): https://github.com/jolie/jolie/blob/master/test/library/json_utils.ol

**Comment by klag** [comment]
Matthias is right. In Jolie a void is transformed into a null in JSON. I think it is the good default choice because there is a native correspondance between two types.

We could mitigate this issue by introducing a special subfield key which allows to transform a void into an empty object.

if  

a.c = Void  (where Void in a Void variable)

is transformed into
{ "a":{ "c":null }}

a.c._EMPTYOBJECT = Void (or something similar, we could discuss the syntax) 

could be transformed into
{ "a":{ "c":{} }}

This means that everytime you need an empty object {} you need to place the field _EMPTYOBJECT (or whatever string we decide) in the target subnode.

What do you think?

**Comment by fmontesi** [comment]
Mmh, a tough one... there's no good answer to whether a Jolie `void` should be a `null` or `{}`. We chose `null` because it seems to be used a lot. But `null` does not exist in Jolie, and `void` is actually `{}` (empty value/object). If you look at how typing works in Jolie and other languages, it is quickly revealed that Jolie has no `null`, since `null` is something that implements any object type but `void` in Jolie doesn't. If you're familiar with more advanced type systems, `void` is more like `Unit` (the fact that we use nulls in the Java implementation is only an internal optimisation, since Java knows no better). But translating `void` to `{}` wouldn't solve this problem anyway, because then we would have the same issue with losing `null` around in JSONs.

In XML we already used a solution similar to what Claudio proposed, for attributes ("@Attributes"). But I don't know if JSON has any sort of "prohibited" node names. For example, in Claudio's suggestion, I wouldn't know whether we should output:

```
{ "a":{ "c":{ "_EMPTYOBJECT": null }}
```

What we could do, however, is to pass the name of the "special node" as a parameter to JsonUtils, so that the programmer can manually decide what's safe. This may lead to very hairy situations in programs, but so far I see no better option.

Another solution we adopt in handling XML is to ask for an XML schema, and use that to know what we should output. Could we do that here too? http://json-schema.org/

**Comment by jacopoMauro** [comment]
As a user, what is important for me is that if I convert the string into a JSON object and then the JSON object into a string I will get the same string I was starting with.
Whatever solution will be able to achieve that will be fine.

**Comment by klag** [comment]
The problem here is to have a valid transformation in both directions.
If A is a JSON string and B a value tree in jolie, we should have:

```
jsonToValue( A ) = B
valueToJson( B ) = A
```

and viceversa

Even if at the beginning I though it was better to keep the transformation void <-> null, now I changed my idea and I think that, maybe, the right solution is to have 
`void -> {}`
and `"$JSON_NULL" -> null `  (where "$JSON_NULL" is a constant string which represents the null value in a jolie tree)  

This solution should allow for a more intuitive manipulation of the value tree into the behaviour of a jolie service. Indeed, the "$JSON_NULL" is a value that can be easy tested in the code with a simple if statement:
`if ( a.b == "$JSON_NULL" ) { .... }`
and the empty node {} is tested by checking the void type:
`if ( a.b instanceof void ) `
They are very intuitive and there is a good correspondance with their real value

On the contrary, if we adopt my previous suggestion (_EMPTYOBJECT) the test on the empty object will become:
`if ( is_defined( a.b._EMPTYOBJECT ) )  { ... } `

Moreover, as far as the assignments are concerned, we should have
`a.b = "$JSON_NULL"  `  //  for null assignment
`a.b = Void   `   // for {} assignment (where Void is a not initialized variable) 

If we follow the other solution we have:
`a.b = Void `// for null assignment
`a.b._EMPTYOBJECT = "" `// here the value of the node _EMPTYOBJECT does not matter because the information is in the node name.

The counterpart is that it is impossible to map the json string` { a:"$JSON_NULL"}`
But, as Fabrizio suggested for the node _EMPTYOBJECT, we could give the possibility to change the constant string "$JSON_NULL" in the configuration of the protocol by making it parameterizable.


What do you think?

**Comment by jacopoMauro** [comment]
Works for me, assuming that the choice is well documented.

**Comment by fmontesi** [comment]
Mmh, we still have some work to do. I agree that using a "null" node (`_EMPTYOBJECT`) would pollute the data structure. But unfortunately, so would `"$JSON_NULL"` when interacting with another protocol!

Imagine:

```
jsonToValue(A) = B  // say that A contains a null
valueToXML(B) = C // now C contains a weird "$JSON_NULL" string instead of a nil XML node
```

So at the very least we need to adopt a consistent way of dealing with "null" for *all* of our transformers.

Looks like we're not alone fighting this:

https://stackoverflow.com/questions/21120999/representing-null-in-json
https://github.com/apiaryio/mson/issues/26

Looks to me like the best people came up with are either:
- Provide the JSON transformers with a type (schema) to tell the transformer whether it should use null or an empty object. This is clean (avoids clashes with possibly used strings, like "$JSON_NULL"), but you might have that valueToJson and jsonToValue are not the inverse of one another.
- Establish a convention for what to do, i.e., an order of preference for when to use `null` or `{}` (see the Stack Overflow discussion). This is simpler (no need for schemas), but less reliable: if you interact with something that uses a different convention, you risk breaking.

I'm worried about users switching from one data format to another and having bad surprises. Maybe we should have a clearly documented _subset_ of each data format that says when you can get in trouble. In general I would like that users don't need to think about this unless they absolutely need to.

Will think about this a bit more... groan.

**Comment by fmontesi** [comment]
OK, here's a typical Gordian Knot solution: let's pass a parameter to JsonUtils that tells whether we should write `{}` (the default in case of no parameter specified) or `null` for void values with no children.

@jacopoMauro Does that smell OK for your examples?
@jolie/developers What do you think?

**Comment by klag** [comment]
But in this case, it will be applied to every node. Maybe, we would like to have it only for some nodes.

**Comment by fmontesi** [comment]
@xpicox I wonder if the latest modification to `isDefined` and `is_defined` might actually fix this... thoughts? Maybe it works already in `master`? :)

**Comment by fmontesi** [comment]
No, nevermind, unfortunately it doesn't. :-\ We still have the problem that an "empty" but defined subnode can be encoded to either `null` or `{}`.
